godot-macros/src/docs.rs

```diff
@@ -0,0 +1,179 @@
+    let members = members
+        .into_iter()
+        .filter(|x| x.var.is_some() | x.export.is_some())
+        .map(member)
+        .collect::<String>();
+    quote::quote![[#base, #desc, #members]]
+}
+
+pub fn inherent_impl(
+    functions: impl IntoIterator<Item = FuncDefinition>,
+    constants: impl IntoIterator<Item = ConstDefinition>,
+    signals: impl IntoIterator<Item = SignalDefinition>,
+) -> TokenStream {
+    let methods = functions.into_iter().map(method).collect::<String>();
+    let signals = signals.into_iter().map(signal).collect::<String>();
+    let constants = constants
+        .into_iter()
+        .map(|ConstDefinition { raw_constant: x }| x)
+        .map(constant)
+        .collect::<String>();
+    quote::quote![[#methods, #signals, #constants]]
+}
+
+pub fn virtual_impl(imethods: impl IntoIterator<Item = Function>) -> String {
+    imethods.into_iter().map(imethod).collect::<String>()
+}
+
+/// `///` is expanded to `#[doc = "…"]`.
+/// this function goes through and extracts the … and converts it to bbcode
+/// for godot's consumption.
+fn docs(doc: impl IntoIterator<Item = Attribute>) -> String {
+    let md = doc
+        .into_iter()
+        .filter_map(|x| {
+            let attr = x.path.first()?;
+            // find #[doc]
+            (attr.to_string() == "doc").then(|| {
+                match x.value {
+                    // #[doc = "…"]
+                    AttributeValue::Equals(_, doc) => Some(
+                        doc.into_iter()
+                            .map(|x| match x {
+                                // venial??
+                                proc_macro2::TokenTree::Literal(x) => x
+                                    .to_string()
+                                    .trim_start_matches('"')
+                                    .trim_end_matches('"')
+                                    .to_string(),
+                                _ => unreachable!(),
+                            })
+                            .reduce(|a, b| format!("{a} {b}"))
+                            .unwrap(),
+                    ),
+                    AttributeValue::Group(_, _) => None,
+                    // valid forms of #[doc] are
+                    // #[doc(hidden|inline|...)] or
+                    // #[doc = string]
+                    _ => unreachable!(),
+                }
+            })?
+        })
+        .collect::<String>();
+    md2bb::convert(&md)
+}
+
+fn signal(signal: SignalDefinition) -> String {
+    let name = signal.signature.name;
+    let params = params(signal.signature.params.iter().filter_map(|(x, _)| match x {
+        FnParam::Receiver(_) => None,
+        FnParam::Typed(y) => Some((y.name.clone(), y.ty.clone())),
+    }));
+    let desc = docs(signal.external_attributes);
+    format!(
+        r#"
+<signal name="{name}">
+  {params}
+  <description>
+  {desc}
+  </description>
+</signal>
+"#
+    )
+}
+
+fn constant(cnst: Constant) -> String {
```

Comment:
you can just name this variable `constant`, i think that's more readable than `cnst`